MRG: Refactor Epochs classes

Eric89GXL opened the pull request:
Not working yet. I hope to keep this more red than green. To-do list:

* [x] Move methods to `_BaseEpochs`
* [x] Make `EpochsArray` subclass `_BaseEpochs` instead of `Epochs`
* [x] Refactor how projection is done (move to `_BaseEpochs` constructor; #2200)
* [x] Refactor how decimation is done (add method, call it in constructor; #2178)
* [x] Refactor how rejection/dropping is done (add method; #1893)
* [x] Get EpochsArray working again
* [x] Fix RtEpochs
* [x] Fix EpochsKit, make it subclass `_BaseEpochs` instead of `EpochsArray` (which in turn had been subclassing `Epochs`)
* [x] Add stronger docstring note about raw resampling / epoch decimation.
* [x] Add test of proj (#2200)
* [x] Add tests of EpochsArray mentioned in #1963
* [x] Add tests of repeated decimation (#2178)
* [x] Add tests of repeated dropping (#1893)

Closes #2225.
Closes #2200.
Closes #2178.
Closes #1982.
Closes #1893.

mainakjas commented:
@Eric89GXL is the post-hoc rejection ready to be tested? I can try it on my data if so

mainakjas commented on mne/epochs.py:
```diff
@@ -1573,7 +1363,207 @@ def equalize_event_counts(self, event_ids, method='mintime', copy=True):
+        'NO DATA' or 'TOO SHORT' if epoch didn't contain enough data;
+        names of channels that exceeded the amplitude threshold;
+        'EQUALIZED_COUNTS' (see equalize_event_counts);
+        or user-defined reasons (see drop_epochs).
```
also 'USER' I think? if it's dropped interactively?

Eric89GXL replied:
this was copy-pasted, but I can look to see if it's documented elsewhere and try to make it consistent

Eric89GXL commented:
@mainakjas I have no idea if it will work because I haven't even tried it yet (just put the framework in place), but you're welcome to if you're bored :)

Eric89GXL commented:
Alright, everything seems to be working at my end. For the record, there is more green than red here only because I added a bunch of testing lines :)

@mainakjas feel free to try the repeated rejection. I added some tests so hopefully it works. You can also try repeated decimation if you want. It would be good to get plenty of eyes on this since it touches a lot of stuff, so whoever has time @dengemann @agramfort @teonlamont please also have a look.

agramfort commented on mne/epochs.py:
```diff
@@ -1573,7 +1382,207 @@ def equalize_event_counts(self, event_ids, method='mintime', copy=True):
+    info: dict
+        Measurement info.
+    event_id : dict
+        Names of  of conditions corresponding to event_ids.
```
of of -> of

dengemann commented on mne/epochs.py:
```diff
@@ -39,39 +37,87 @@
+            self.events = events
+            del events
+        else:
+            self.drop_log = []
```
list()

teonlamont commented on mne/epochs.py:
```diff
@@ -121,54 +167,138 @@ def __init__(self, info, event_id, tmin, tmax, baseline=(None, 0),
+            epochs._decim_slice = slice(None, None, None)
+            epochs._decim = 1
+            epochs.times = epochs._raw_times
+        else:
```
Does this mean you can decimate your data before preloading, e.g. `epochs.decimate(4).preload_data()`?

Eric89GXL replied:
Yes

teonlamont commented on mne/epochs.py:
```diff
@@ -1573,7 +1382,207 @@ def equalize_event_counts(self, event_ids, method='mintime', copy=True):
+        by event_id, they will be marked as 'IGNORED' in the drop log.
+    event_id : int | list of int | dict | None
+        The id of the event to consider. If dict,
+        the keys can later be used to acces associated events. Example:
```
acces -> access

teonlamont commented on mne/epochs.py:
```diff
@@ -1573,7 +1382,207 @@ def equalize_event_counts(self, event_ids, method='mintime', copy=True):
+    picks : array-like of int | None (default)
+        Indices of channels to include (if None, all channels are used).
+    name : string
+        Comment that describes the Evoked data created.
```
Evoked --> Epochs

teonlamont commented on mne/epochs.py:
```diff
@@ -1573,7 +1382,207 @@ def equalize_event_counts(self, event_ids, method='mintime', copy=True):
+    event_id : dict
+        Names of  of conditions corresponding to event_ids.
+    ch_names : list of string
+        List of channels' names.
```
suggestion: channels' names --> channel names

teonlamont commented on mne/tests/test_epochs.py:
```diff
@@ -61,14 +60,144 @@ def _get_data():
+    """Test epochs rejection
+    """
+    raw, events, picks = _get_data()
+    # cull the list just to contain the relevent event
```
relevent --> relevant

teonlamont commented on mne/epochs.py:
```diff
@@ -946,8 +792,13 @@ def drop_bad_epochs(self):
             disk. To avoid reading epochs form disk multiple times, initialize
```
form --> from

teonlamont commented:
if you save an epochs after baseline correction, load it, then decimate it, it's baseline-correcting again, should this be the behavior?

teonlamont commented:
wait, I take that back, it's happening with the `apply_proj`, not `decimate`.

Eric89GXL commented:
I don't think multiple baseline corrections should hurt things. It might be a little bit better not to, but I think it will further increase code complexity to try to keep track of whether or not baseline correction has already been done or not.

In any case, I think this should actually also be happening in `master`, so if it is a problem, we should tackle it in a separate PR to avoid making this one more complex.

Eric89GXL commented:
Thanks for the quick reviews. Comments thus far addressed.

agramfort commented:
besides the API discussion about the way to reject using param in drop_bad_epochs LGTM

doc builds fine !

agramfort commented:
@mainakjas can you see with your data if it fixes your issue?

Eric89GXL commented:
@agramfort done, now they are set in `drop_bad_epochs`, see latest commit.

teonlamont commented:
so we have both `epochs.drop_epochs` and `epochs.drop_bad_epochs`. they have similar goals. should we consider consolidating the functions or would that create too much a burden?

teonlamont commented:
i've tested it out and it seems to me working fine for me. other than my last comment, which does not directly affect this pr, and travis being unhappy, LGTM. great improvements to Epochs under the hood :)

mainakjas commented:
I don't have access to the data because our office is shifting and the data is on the hard drive :(

mainakjas commented on mne/tests/test_epochs.py:
```diff
@@ -61,14 +60,151 @@ def _get_data():
+        assert_equal(len(epochs), len(events))
+        epochs.drop_bad_epochs(reject)
+        assert_equal(len(epochs), len(events) - 4)
+        assert_array_equal(epochs.selection, selection)
```
there are no tests to check if the data remains the same?

Eric89GXL replied:
Yikes, good call! I've now added the tests, they uncovered a bug, and also caused me to simplify (I hope) the `_get_data` function.

Eric89GXL commented:
@teonlamont I think of the functions as doing slightly different things, one is automated rejection the other is manual rejection/subselection. The naming is probably closer than it would ideally be, but deprecation would probably be too much of a pain.

Also, one problem with combining them is that `drop_epochs` and `drop_bad_epochs` both modify the selection. The order in which they occur thus actually matters. In that sense it's probably better to have them be in two different functions so the order is explicit in people's code.

Eric89GXL commented:
@mainakjas comments addressed. Thanks for the quick reviews y'all.

mainakjas commented on mne/tests/test_epochs.py:
```diff
@@ -61,14 +60,160 @@ def _get_data():
+        assert_equal(len(epochs), len(epochs.get_data()))
+        assert_array_equal(epochs.selection, selection)
+        assert_array_equal(epochs.drop_log, drop_log)
+        print(preload)
```
... note to remove it. Debugging artifact I suppose

Eric89GXL replied:
removed, thanks

mainakjas commented on mne/tests/test_epochs.py:
```diff
@@ -61,14 +60,160 @@ def _get_data():
+
+    data_7 = None
+    keep_idx = [0, 1, 2]
+    for preload in (True, False):
```
hmm ... have you tried this with `proj=True` as well?

Eric89GXL replied:
added in latest commit (thankfully passed)

mainakjas commented on mne/tests/test_epochs.py:
```diff
@@ -61,14 +60,162 @@ def _get_data():
+            assert_array_equal(epochs.get_data(), data_7[proj][keep_idx])
+
+            # rejection twice
+            reject_part = dict(grad=1100e-12, mag=4e-12, eeg=80e-6, eog=150e-6)
```
what happens if I set a `reject` param again through `drop_bad_epochs`? Is it allowed?

Eric89GXL replied:
Yes, that gets tested somewhere in here, no?

mainakjas replied:
hmm ... I think if you apply a lenient threshold followed by a strict threshold, it should give the same result as applying the strict threshold alone. But I don't see that test anywhere. Notwithstanding that, I am not sure if it's even wise to allow that. Because you won't have a log of the thresholds applied anywhere.

Eric89GXL replied:
That is actually tested on the lines we are commenting on... `reject_part` is more lenient and only gets rid of one epoch, whereas `reject` gets rid of four of them.

I guess the question is, how should this function? Currently, the functionality is that you can pass `reject` and `flat` as many times as you want to `drop_bad_epochs`, and it will drop any existing epochs that meet those criteria.

An alternative would be to make it so that you can only do `drop_bad_epochs` with a non-`None` `reject` and `flat` exactly once. Is that what you want? That would allow you to create epochs with `reject = flat = None` with any state of `proj` and `preload`, and then at some point later apply any rejection criteria you want.

mainakjas replied:
On Sun, Jun 21, 2015 at 1:51 AM, Eric Larson <notifications@github.com>
wrote:

> In mne/tests/test_epochs.py
> <https://github.com/mne-tools/mne-python/pull/2211#discussion_r32889151>:
>
> > +
> > +            # rejection post-hoc
> > +            epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
> > +                            preload=preload)
> > +            epochs.drop_bad_epochs()
> > +            assert_equal(len(epochs), len(events))
> > +            assert_array_equal(epochs.get_data(), data_7[proj])
> > +            epochs.drop_bad_epochs(reject)
> > +            assert_equal(len(epochs), len(events) - 4)
> > +            assert_equal(len(epochs), len(epochs.get_data()))
> > +            assert_array_equal(epochs.selection, selection)
> > +            assert_array_equal(epochs.drop_log, drop_log)
> > +            assert_array_equal(epochs.get_data(), data_7[proj][keep_idx])
> > +
> > +            # rejection twice
> > +            reject_part = dict(grad=1100e-12, mag=4e-12, eeg=80e-6, eog=150e-6)
>
> That is actually tested on the lines we are commenting on... reject_part
> is more lenient and only gets rid of one epoch, whereas reject gets rid
> of four of them.
>
> I guess the question is, how should this function? Currently, the
> functionality is that you can pass reject and flat as many times as you
> want to drop_bad_epochs, and it will drop any existing epochs that meet
> those criteria.
>
> An alternative would be to make it so that you can only do drop_bad_epochs
> with a non-None reject and flat exactly once. Is that what you want?
>
yeah I think this is exactly what @agramfort wanted. Any other opinions?
@teonlamont?

> That would allow you to create epochs with reject = flat = None with any
> state of proj and preload, and then at some point later apply any
> rejection criteria you want.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2211/files#r32889151>.
>

Eric89GXL replied:
I'm fine with either way of doing it. I'll wait to hear from others (@teonlamont @agramfort @dengemann) before changing it.

Even in the new scheme I think we'll still need an `'existing'` or `'original'` argument in the new scheme, in order to differentiate it from `None`, which means "no rejection of this type".

teonlamont replied:
I don't see a problem with being able to run the drop_bad_epochs multiple times. even if it's superfluous, it shouldn't be problematic. typical case: reject once with lenient dict(mag=4e-12) but waveform is crappy, then a more stringent dict(mag=3e-12), it does a better job.

mainakjas commented on mne/tests/test_epochs.py:
```diff
@@ -61,14 +60,162 @@ def _get_data():
+    data_7 = dict()
+    keep_idx = [0, 1, 2]
+    for preload in (True, False):
+        for proj in (True, False):
```
and I suppose `proj='delayed'` throws an error?

Eric89GXL replied:
Haven't tried, will add in a bit.

Eric89GXL replied:
delayed works, added in latest commit

mainakjas commented on mne/epochs.py:
```diff
@@ -682,272 +799,47 @@ def plot_psd_topomap(self, bands=None, vmin=None, vmax=None, proj=False,
-            self._data = None
-
-    def drop_bad_epochs(self):
+    def drop_bad_epochs(self, reject='original', flat='original'):
```
but allowing a string here is inconsistent with what you do in the `Epochs` constructor, no?

Eric89GXL replied:
Yes but that's because there, there is no existing one. I could change it
to 'existing'?

mainakjas commented on mne/epochs.py:
```diff
@@ -682,272 +799,47 @@ def plot_psd_topomap(self, bands=None, vmin=None, vmax=None, proj=False,
+        Dropping bad epochs can be done multiple times with different
+        ``reject`` and ``flat`` parameters. However, once an epoch is
+        dropped, it is dropped forever, so if more lenient thresholds may
+        subsequently be applied, `epochs.copy` should be used.
```
dangerous to do this, no? We should allow setting `reject` only once, no?

Eric89GXL replied:
I thought the whole point was to allow doing rejection multiple times...

mainakjas replied:
yeah, but you can set the rejection parameter only once. You can do it multiple times just to make sure that the epochs have been dropped. That's the only reason.

Eric89GXL replied:
> You can do it multiple times just to make sure that the epochs have been dropped

How do you mean? You mean re-do the work of reading the data and checking the thresholds (even if nothing has been changed) to make sure the function works properly and returns the same result? I'm -1 on that. It seems like you'd want the first call to do something, and any subsequent call to short-circuit and assume the first call worked (e.g., for efficiency). I don't think it's a good idea to have to code things inefficiently so that users can double-check our work -- I think it would be better just to get it right the first time as best we can. After all, extending this type of idea, if I understand it correctly, would mean taking functions like `get_data` and having them re-read all the data from disk every time you called them, which would not be good...?

mainakjas replied:
no no, not rechecking the thresholds. Short-circuiting is the way to go. But the user may want to use it a couple of times in their script just to make sure that they have dropped the epochs (in case they are not sure it was done in some other part of the script). But you want to keep the same thresholds, not change it. I think we agree?

Eric89GXL replied:
But is it supposed to also have the functionality I mention above, namely allowing instantiating with `reject=flat=None`, and then at some later point applying some criteria? I guess I'm still not clear if the goal is just to 1) sanity check values like you say, or also 2) allow instantiating with `None` and doing rejection later.

I had assumed that if a user wants to know if they have rejected, they can compare `epochs.reject` to whatever they expect their rejection threshold to be. Is the idea to have `drop_bad_epochs` do that for them, then? If we don't want the instantiate-with-None-reject-later-with-something functionality mentioned above, then it wouldn't make sense to add arguments to `drop_bad_epochs`. 

FWIW the version I have programmed here essentially does the same thing that @dengemann's gist does from #1893.

I'm also confused because the re-calling thing seems to be new, and not what was discussed in #1900 (@mainakjas followed by @agramfort comments):

>>  so we throw an error also in the case that the user calls drop_bad_epochs
>> with the same reject and flat parameters?
>
> yes. It should not happen in real life for me

So it sounds like from before that dropping should only be allowed exactly once, and that subsequent calls to `drop_bad_epochs` with any arguments should be an error. What you are arguing for is different now, right, in that subsequent calls should throw an error only if the arguments don't match?

mainakjas replied:
On Sun, Jun 21, 2015 at 2:19 AM, Eric Larson <notifications@github.com>
wrote:

> In mne/epochs.py
> <https://github.com/mne-tools/mne-python/pull/2211#discussion_r32889287>:
>
> > +            Valid keys are 'grad' | 'mag' | 'eeg' | 'eog' | 'ecg'.
> > +            If reject is None then no rejection is done. If 'original',
> > +            then the rejection parameters set at instantiation are used.
> > +        flat : dict | str | None
> > +            Rejection parameters based on flatness of signal.
> > +            Valid keys are 'grad' | 'mag' | 'eeg' | 'eog' | 'ecg', and values
> > +            are floats that set the minimum acceptable peak-to-peak amplitude.
> > +            If flat is None then no rejection is done. If 'original',
> > +            then the flat parameters set at instantiation are used.
> > +
> > +        Notes
> > +        -----
> > +        Dropping bad epochs can be done multiple times with different
> > +        ``reject`` and ``flat`` parameters. However, once an epoch is
> > +        dropped, it is dropped forever, so if more lenient thresholds may
> > +        subsequently be applied, `epochs.copy` should be used.
>
> But is it supposed to also have the functionality I mention above, namely
> allowing instantiating with reject=flat=None, and then at some later
> point applying some criteria?
>
yes, it is supposed to have that. But once you apply it. Any subsequent
calls should basically short-circuit. But you should allow subsequent-calls
with the same parameter I think.

> I guess I'm still not clear if the goal is just to 1) sanity check values
> like you say, or also 2) allow instantiating with None and doing
> rejection later.
>
I'm not sure what you mean by 1/. 2/ definitely. And also just allow
calling the function as many times with the same reject parameters as
before (or if it's None in the function call -- check using
epochs.is_bad_dropped and short circuit).

> I had assumed that if a user wants to know if they have rejected, they can
> compare epochs.reject to whatever they expect their rejection threshold
> to be.
>
yeah, but how does this tell you that it has been rejected when
preload=False?

> Is the idea to have drop_bad_epochs do that for them, then?
>
yep, it should just force a rejection if it's not been done yet. If it's
been done, short-circuit as you say.

> If we don't want the instantiate-with-None-reject-later-with-something
> functionality mentioned above, then it wouldn't make sense to add arguments
> to drop_bad_epochs.
>
We do want it I think

> FWIW the version I have programmed here essentially does the same thing
> that @dengemann <https://github.com/dengemann>'s gist does from #1893
> <https://github.com/mne-tools/mne-python/issues/1893>. I'm also confused
> because the re-calling thing seems to be new, and not what was discussed in
> #1900 <https://github.com/mne-tools/mne-python/pull/1900> (@mainakjas
> <https://github.com/mainakjas> followed by @agramfort
> <https://github.com/agramfort> comments):
>
 so we throw an error also in the case that the user calls drop_bad_epochs
> with the same reject and flat parameters?
>
> yes. It should not happen in real life for me
>
> So it sounds like from before that dropping should only be allowed exactly
> once, and that subsequent calls to drop_bad_epochs with any arguments
> should be an error. What you are arguing for is different now, right, in
> that subsequent calls should throw an error only if the arguments don't
> match?
>
If you can make it not throw an error, that would be great! I think I was
having some implementation troubles. Also, the way we did it was to have
drop_bad_epochs() take the None arguments for both the parameters if we
wanted to keep the same reject as before. I think ...

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2211/files#r32889287>.
>

dengemann replied:
does not return self, inconsistent

Eric89GXL commented:
Pulling discussion to the main thread for simplicity.

The way I see it, there are three options on the table:

1. Allow Epochs to be instantiated with any valid `reject` and `flat` arguments, and `drop_bad_epochs` to be called multiple times with any valid `reject` and `flat` arguments, doing rejection of any remaining epochs using the parameters when `drop_bad_epochs` gets called. This was @dengemann's gist from #1893 and what is done here.

2. Allow rejection to be done using parameters that actually cause rejection exactly once. Thus "meaningful" arguments are allowed for `drop_bad_epochs` if and only if the class was instantiated with `reject=flat=None`. `drop_bad_epochs` (and `__init__`) can only be called once with non-`None` arguments. This was what @agramfort argued for I think.

3. Same as 2), but allow multiple calls to `drop_bad_epochs` to be called multiple times, so long as the `reject` and `flat` parameters match each time. I think this is what @mainakjas you are arguing for.

I actually personally prefer (2) to (3) or (1). I think (2) forces you to be more careful with your coding. If we want people to know whether or not bad has been dropped, I'd rather have a `is_bad_dropped` argument, which can be combined with `reject` and `flat` to determine what the status is. But I can live with 1) or 3) as people prefer.

FWIW, the way I have things coded, the `preload` option does not affect the outcome of the dropping, and it won't matter in any of these cases so far as I can tell. Bads have either been dropped by `drop_bad_epochs` explicitly in the `preload=False` case, or by `preload_data()` implicitly in the `preload=True` case. In other words, with `preload=False`, you'd be able to tell the state by looking at `epochs.reject`, `epochs.flat`, and `epochs.bad_dropped` (if we make that param public instead of private, which makes sense to me). If `epochs.reject is not None and epochs.bad_dropped`, then bad epochs have been dropped already, regardless of `preload` state.

I am -1 on making `None` parameters for `reject` and `flat` in `drop_bad_epochs` mean "do whatever was instantiated". That changes the meaning of `None` for `reject` and `flat` between the constructor (and everywhere else IIRC), where it means "do no rejection of this type", and the `drop_bad_epochs` method, where it would instead mean "do whatever was instantiated".

Eric89GXL commented on mne/io/pick.py:
```diff
@@ -188,6 +188,10 @@ def pick_types(info, meg=True, eeg=False, stim=False, eog=False, ecg=False,
+    from .meas_info import Info
+    if not isinstance(info, Info):
+        raise TypeError('info must be an instance of Info, not %s'
+                        % type(info))
```
FYI I made this change because we had no check to ensure `info` was an instance of `Info` here before, and I got some weird error when I accidentally passed `raw` instead of `raw.info`. We could relax this check to `isinstance(info, (Info, dict))` but I'd prefer to keep it more strict to `Info` for now to help ensure people create their info structs correctly by reading or by using `create_info`. This necessitated a couple of minor changes elsewhere to comply with this check.

Eric89GXL commented:
BTW at some point it should also be possible to make an `Epochs` class that reads an `-epo.fif` file from disk as necessary, i.e. adding a `preload=False` option to `read_epochs`. That, coupled with an option to write epochs out without storing all the epochs in memory would reduce the memory burden for people. But we can add that later :)

Eric89GXL commented on mne/forward/_make_forward.py:
```diff
@@ -436,7 +437,7 @@ def make_forward_solution(info, trans, src, bem, fname=None, meg=True,
     # make a new dict with the relevant information
     mri_id = dict(machid=np.zeros(2, np.int32), version=0, secs=0, usecs=0)
-    info = dict(nchan=info['nchan'], chs=info['chs'], comps=info['comps'],
+    info = Info(nchan=info['nchan'], chs=info['chs'], comps=info['comps'],
```
See here @teonlamont

Eric89GXL commented:
So now we have a vote for 1) from @teonlamont and for 3) for @mainakjas.

@dengemann @agramfort let us know what you think.

mainakjas commented:
Does 2/ allow calling the function multiple times with `None` arguments? What will that do?

Eric89GXL commented:
No, not multiple times. And I would argue against allowing None to mean
anything other than "don't reject". If we allow calling with None to be
done in 2 or 3, then they become very close to 1...

mainakjas commented:
okay, but that breaks how the function is being used currently, no?

Eric89GXL commented:
No, 1 is implemented now. So you can do rejection, then call
drop_bad_epochs with two None arguments just fine (it will do nothing, but
it will work).

Eric89GXL commented:
Or by "now" do you mean in master? Yes 2 would break existing code that
called drop_bad_epochs multiple times.

mainakjas commented:
Yes, by "now" I meant master :) Exactly, 2/ would break a lot of code in master

agramfort commented:
I actually argued for 2 and I understand the concern about reject=None having
an unfortunate meaning to avoid breaking current API which is to be able to
drop epochs multiple times. The reason for this btw is that some function
start by calling this method (eg. in beamformers) and you don't want it to crash
is the used called it before manually.

I don't think now of a good default value for reject to avoid breaking API...

Eric89GXL commented:
So does that mean you now prefer option 3?
On Jun 21, 2015 1:11 PM, "Alexandre Gramfort" <notifications@github.com>
wrote:

> I actually argued for 2 and I understand the concern about reject=None
> having
> an unfortunate meaning to avoid breaking current API which is to be able to
> drop epochs multiple times. The reason for this btw is that some function
> start by calling this method (eg. in beamformers) and you don't want it to
> crash
> is the used called it before manually.
>
> I don't think now of a good default value for reject to avoid breaking
> API...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2211#issuecomment-113951138>
> .
>

agramfort commented:
How would 3 work if reject is not None in init and drop_bad_epochs is ran multiple times with default params?

Eric89GXL commented:
it would drop bad epochs exactly once (during init if `preload=True`, otherwise on the first call to `drop_bad_epochs(reject='existing', flat='existing')`) and then on every subsequent call to `drop_bad_epochs(reject='existing', flat='existing')` or `drop_bad_epochs(reject=reject, flat=flat)` where `reject` and `flat` are identical to those passed during init, will basically `pass` -- it would detect that dropping has already been done and return.

agramfort commented:
> existing is the good idea I was missing. 
> 
Sounds like a plan to me

Eric89GXL commented:
@dengemann you okay with option 3 (only allow rejection to actually occur once) instead of option 1 (what you had implemented; arbitrary re-rejections of remaining epochs)? I think you can get the behavior of 3) if you want it using 1) through use of `epochs.copy()` instances, albeit with worse memory use.

agramfort commented:
after a good night of sleep, I am thinking that with the existing option
there is no strong reason to prevent people from changing the reject
parameter multiple times as long as the values are decreasing for each
channel type for reject and increasing for flat. That way if you selected a
too high eog param as first guess you can try to decrease it a bit to see
what happens.

teonlamont commented:
+1 with @agramfort. I have the same opinion on the matter.

mainakjas commented:
yeah, I think I like @agramfort 's idea

teonlamont commented:
this is related and might be addressed when we change the reject but currently, if `epochs.reject == None` and you set `epochs.reject = X`, then `epochs.drop_bad_epochs()`, nothing happens.

Eric89GXL commented:
@teonlamont that's the way I had it before, you would set `epochs.reject` and then do `epochs.drop_bad_epochs` but @agramfort made me change it :) I could make `epochs.reject` a read-only property, but dicts are mutable so a person might think that `epochs.reject['eeg'] = 10e-6` would actually do something meaningful when it would not. Maybe we should go back to the old syntax...

Eric89GXL commented:
At least it sounds like we've settled on a modified version of point (1) for the implementation, so I'll implement that bit.

agramfort commented:
>
> this is related and might be addressed when we change the reject but
> currently, if epochs.reject == None and you set epochs.reject = X, then
> epochs.drop_bad_epochs(), nothing happens.
>

I don't like allowing people to tweak manually attributes. It's error prone
as we cannot easily do checks.

Eric89GXL commented:
FYI you can make things properties with a getter and setter to do checks, but in this case it would not solve the mutability issue

teonlamont commented:
maybe add the reject state to the `__repr__` and have reject private or not there. I would fall into the trap of thinking I did the right thing in that case. I think it would be the safest to have the reject only be settable via the method drop_bad_epochs or the constructor.

Eric89GXL commented:
@teonlamont the reject state is already in the `__repr__` as `(all good)` or `(good and bad)`

Eric89GXL commented:
I will make `reject` and `flat` read-only, though

Eric89GXL commented:
(other than changes like `epochs.reject['eeg'] = ...`, which we can't easily do anything about)

agramfort commented:
don't do black magic. We never modify attributes manually in examples.
People who do take the risk.

Eric89GXL commented:
So you don't want me to add this?
```
@property
def reject(self):
    return self._reject.copy()
```
This would make it so that someone doing `self.reject = ...` would at least get an error message. It doesn't seem too magical to me?

Eric89GXL commented:
@agramfort added an increasing-stringency test, ready for review/merge from my end.

agramfort commented:
besides +1 for merge.

Eric89GXL commented:
@agramfort done. @mainakjas can you have a last look and merge if you're happy?

agramfort commented:
update what's new to document fixes and ability to reject iteratively?

mainakjas commented:
@Eric89GXL I'll take a last look tomorrow morning and press the green button if everything works.

dengemann commented on mne/epochs.py:
```diff
@@ -39,39 +37,87 @@
         else:
             raise ValueError('event_id must be dict or int.')
+        self.event_id = event_id
+        del event_id
```
why is this needed?

dengemann commented on mne/epochs.py:
```diff
@@ -39,39 +37,87 @@
+            else:
+                raise ValueError('No desired events found.')
+            self.events = events
+            del events
```
why delete events? no copies involved, etc.

Eric89GXL replied:
It's a safety thing -- by removing it from the namespace we ensure that
from that point forward we only make use of self.events

dengemann commented on mne/io/meas_info.py:
```diff
@@ -51,6 +51,16 @@ class Info(dict):
+        info : instance of Info
+            The copied info.
+        """
+        return Info(super(Info, self).copy())
```
mhmm. wondering where we use this. It will be a shallow copy, wondering if we really wan this,

mainakjas commented:
@Eric89GXL rejection works like a charm on my data!

mainakjas commented:
Let me check if `RtEpochs` is intact ...

mainakjas commented:
Actually, I get this error @Eric89GXL: 

```py
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tsi/doctorants/mainak/xyz.py in <module>()
    129     fix_stim_artifact(epochs, stim_channel='STI101')
    130     # epochs.drop_bad_epochs(reject=dict(eeg=200e-6))
--> 131     epochs.drop_bad_epochs(reject=dict(grad=4000e-13, mag=4e-12, eog=250e-6))
    132     # epochs.plot(trellis=False)
    133 

/tsi/doctorants/mainak/mne-python/mne/epochs.pyc in drop_bad_epochs(self, reject, flat)
    863             raise ValueError('reject and flat, if strings, must be "existing"')
    864         self._reject_setup(reject, flat)
--> 865         self._get_data(out=False)
    866 
    867     def drop_log_stats(self, ignore=['IGNORED']):

/tsi/doctorants/mainak/mne-python/mne/epochs.pyc in _get_data(self, out, verbose)

/tsi/doctorants/mainak/mne-python/mne/utils.pyc in verbose(function, *args, **kwargs)
    538         finally:
    539             set_log_level(old_level)
--> 540     return function(*args, **kwargs)
    541 
    542 

/tsi/doctorants/mainak/mne-python/mne/epochs.pyc in _get_data(self, out, verbose)
   1070             # adjust the data size if there is a reason to (output or update)
   1071             if out or self.preload:
-> 1072                 data.resize((n_out,) + data.shape[1:], refcheck=False)
   1073 
   1074         return data if out else None

ValueError: cannot resize this array: it does not own its data
```

mainakjas commented:
This is to reproduce:

```py
import mne
from mne import io
from mne.datasets import sample

data_path = sample.data_path()

raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
event_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw-eve.fif'
tmin, tmax = -0.2, 0.5

event_id = {'Auditory/Left': 1, 'Auditory/Right': 2}

raw = io.Raw(raw_fname)
events = mne.read_events(event_fname)

# pick EEG and MEG channels
picks = mne.pick_types(raw.info, meg=False, eeg=True, stim=False, eog=True,
                       exclude='bads')
epochs = mne.Epochs(raw, events, event_id, tmin, tmax, picks=picks,
                    baseline=(None, 0), reject=None,
                    preload=True)
epochs = epochs[:10]
epochs.drop_bad_epochs(reject=dict(eeg=80e-6, eog=150e-6))
```

Eric89GXL commented:
I expect many of the RtEpochs functions not to work actually. Are you
testing it just by using some dummy code and trying all the methods? If you
share that script I can go through and deal with the functions.

mainakjas commented:
This error isn't related to `RtEpochs`. I just shared the code to reproduce. I need to test `RtEpochs`. Basically, if the examples work, I think most of it should work ...

Eric89GXL commented:
Thanks @mainakjas I'll try to fix it today

Eric89GXL commented:
@mainakjas fixed and test added, your code runs fine over here now

agramfort commented:
@mainakjas happy?

mainakjas commented:
@Eric89GXL when I try to run the example `ftclient_rt_compute_psd.py`, I get the error:

```py
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/home/mainak/Desktop/projects/mne-python/examples/realtime/ftclient_rt_compute_psd.py in <module>()
     43     n_samples = 2048  # time window on which to compute FFT
     44     for ii in range(20):
---> 45         epoch = rt_client.get_data_as_epoch(n_samples=n_samples, picks=picks)
     46         psd, freqs = compute_epochs_psd(epoch, fmin=2, fmax=200, n_fft=n_fft)
     47 

/home/mainak/Desktop/projects/mne-python/mne/realtime/fieldtrip_client.py in get_data_as_epoch(self, n_samples, picks)
    258 
    259         # create epoch from data
--> 260         epoch = EpochsArray(data, self.info, events)
    261 
    262         # pick channels

/home/mainak/Desktop/projects/mne-python/mne/epochs.py in __init__(self, data, info, events, tmin, event_id, reject, flat, reject_tmin, reject_tmax, baseline, verbose)

/home/mainak/Desktop/projects/mne-python/mne/utils.py in verbose(function, *args, **kwargs)
    538         finally:
    539             set_log_level(old_level)
--> 540     return function(*args, **kwargs)
    541 
    542 

/home/mainak/Desktop/projects/mne-python/mne/epochs.py in __init__(self, data, info, events, tmin, event_id, reject, flat, reject_tmin, reject_tmax, baseline, verbose)
   1728 
   1729         if len(info['ch_names']) != np.shape(data)[1]:
-> 1730             raise ValueError('Info and data must have same number of '
   1731                              'channels.')
   1732         if data.shape[0] != len(events):

ValueError: Info and data must have same number of channels.

```
This doesn't happen in master

mainakjas commented:
@Eric89GXL I suppose you have the fieldtrip client to try this example? Let me know if not

mainakjas commented:
@Eric89GXL have you tried building the docs? (I haven't fixed my pyface problem yet, so can't try).

mainakjas commented:
Other than this, LGTM. +1 for merge once this is fixed

Eric89GXL commented:
Cool, failing that check suggests there is a hidden bug in master. I don't
have the client on my computer, can you have a look and open a quick pr to
my branch if it's easy? If not I can try to figure out from the code.

Eric89GXL commented:
(I am on windows for a while)

mainakjas commented:
@Eric89GXL I think I know what's wrong. I can take care of it tomorrow morning and make a pr

mainakjas commented:
@Eric89GXL : https://github.com/Eric89GXL/mne-python/pull/19

mainakjas commented:
I have one more comment @Eric89GXL . I get many log messages which are not relevant when creating an `EpochsArray` -- e.g., baselining, proj etc. Can these be suppressed?

agramfort commented:
verbose=False?

mainakjas commented:
no, but I am not saying that there are too many messages. I am saying that those log messages don't make sense when you create an `EpochsArray`.

teonlamont commented:
you can still baseline correct and reject with EpochsArray

agramfort commented:
true but do you see this all the time?

mainakjas commented:
Look at the log for this:

```py
import mne
import numpy as np

sfreq = 1000  # Sampling frequency
times = np.arange(0, 10, 0.001)  # Use 10000 samples (10s)
sin = np.sin(times * 10)  # Multiplied by 10 for shorter phase
cos = np.cos(times * 10)

event_id = 1
events = np.array([[200, 0, event_id],
                   [1200, 0, event_id],
                   [2000, 0, event_id]])  # List of three arbitrary events

epochs_data = [[sin[:700], cos[:700]],
               [sin[1000:1700], cos[1000:1700]],
               [sin[1800:2500], cos[1800:2500]]]

ch_names = ['sin', 'cos']
ch_types = ['mag', 'mag']
info = mne.create_info(ch_names=ch_names, sfreq=sfreq, ch_types=ch_types)

epochs = mne.EpochsArray(epochs_data, info=info, events=events,
                         event_id={'arbitrary': 1})
```

I see
```
3 matching events found
0 projection items activated
No baseline correction applied...
No baseline correction applied...
No baseline correction applied...
0 bad epochs dropped
```
I don't think I'd want to see any of this when I create `EpochsArray`

mainakjas commented:
... or maybe not, I changed my mind. I think it's fine.

agramfort commented:
are we good then?

Eric89GXL commented:
I need to merge his PR into my branch first

agramfort commented:
ok

you have my blessing to merge when you're happy

Eric89GXL commented:
Merged @mainakjas's PR, will merge this branch into `master` assuming the CIs come back happy.

Eric89GXL commented:
Thanks for the reviews and feedback, hope this works well for everybody :)

agramfort commented:
:beers: !!!

For real soon in Paris ;)

mainakjas commented:
Yayy!!!

teonlamont commented:
yay!! thanks @Eric89GXL!
